Fix/3990 buffer emit remainder on complete

benlesh opened the pull request:
### fix(buffer): Remaining buffer will be emited on source close if `closingNotifier` active
    
Gives the author control over the emission of the final buffer. If the `closingNotifier` completes before the source does, no more buffers will be emitted. If the `closingNotifier` is still active when the source completes, then whatever is in the buffer at the time will be emitted from the resulting observable before it completes.
    
BREAKING CHANGE: Final buffered values will now be emitted from the resulting observable if the `closingNotifier` is still active. The simplest workaround if you want the original behavior (where you possibly miss values), is to add a `skipLast(1)` at the end. Otherwise, you can try to complete the `closingNotifier` prior to the completion of the source.
    
Fixes #3990, #6035
    
    
### fix(buffer): closingNotifier completion does not complete resulting observable
    
Resolves an issue where the resulting observable would complete when the closingNotifier completed. Notifier completion should not complete the result, only source completion should do that.
    
BREAKING CHANGE: closingNotifier no longer closes the result of `buffer`. If that is truly a desired behavior, then you should use `takeUntil`. Something like: `source$.pipe(buffer(notifier$), takeUntil(notifier$.pipe(ignoreElements(), endWith(true))))`, where `notifier$` is multicast, although there are many ways to compose this behavior.

benlesh commented on spec/operators/buffer-spec.ts:
```diff
@@ -16,11 +17,25 @@ describe('Observable.prototype.buffer', () => {
       const a = hot('   -a-b-c-d-e-f-g-h-i-|');
       const b = hot('   -----B-----B-----B-|');
-      const expected = '-----x-----y-----z-|';
+      const expected = '-----x-----y-----z-(F|)';
```
What you're going to see through a lot of these tests is I had to add the last emission of an empty buffer for cases when the `closingNotifier` was not complete when the source completed.

cartant replied:
Yeah, but TBH I am a little disappointed that you didn't use an emoji:
```ts
      const a = hot('   -a-b-c-d-e-f-g-h-i-|');
      const b = hot('   -----B-----B-----B-|');
      const expected = '-----x-----y-----z-(🦖|)';
      const expectedValues = {
        x: ['a', 'b', 'c'],
        y: ['d', 'e', 'f'],
        z: ['g', 'h', 'i'],
        '🦖': [],
      };
```

cartant reviewed (approved):
LGTM. I'm going to push another commit to this branch to add some tests that assert/document what I recall you mentioning should be an equivalence between `buffer` and `window` (with `toArray` applied). However, one of the cases in the tests fails. If this equivalence is expected, perhaps we should looks at what's different? I don't have time to do this, right now - I can look at it tomorrow. Feel free to toss the commit if it's misguided or or whatever. I won't be offended. 🙂

benlesh commented:
@cartant it looks like we caught another inconsistency with our API. The `window` operator is allowing the completion of the `windowBoundaries` notifier to act as a notification, rather than a simple completion. It's completing the result *and* the window, when really it should just mean that the author is saying they're not going to close and reopen any new windows again.

(It's a bug)

benlesh commented:
@cartant So I've added another commit that also fixes `window` to behave properly, and changes the behavior of `buffer` to the simpler thing that makes more sense. They are now equivalent. I've also rebased the commits to make sure we don't have conflicting messages in our changelog and added examples on how to get the "old" behavior to the breaking change message.

cartant commented on spec/operators/buffer-spec.ts:
```diff
@@ -16,40 +17,78 @@ describe('Observable.prototype.buffer', () => {
     });
   });
 
+  it('should not emit a final buffer if the closingNotifier is already complete', () => {
```
nitpick: this test description is now incorrect

```
it('should emit ...
```

cartant commented on src/internal/operators/buffer.ts:
```diff
@@ -43,19 +44,38 @@ import { OperatorSubscriber } from './OperatorSubscriber';
  */
 export function buffer<T>(closingNotifier: Observable<any>): OperatorFunction<T, T[]> {
   return operate((source, subscriber) => {
+    // The current buffered values. If this is null, it's because the
+    // closingNotifier has completed before the source.
```
nitpick: I'm not sure about this comment. AFAICT, it's nulled when the `buffer` operator's subscriber is torn down - not the subscriber to the notiffier.

benlesh replied:
Ugh... comment cruft. Thanks for catching that.

cartant reviewed (approved):
LGTM, just some nitpicks.

benlesh commented:
Merging despite `ts@latest` issue that seems unrelated.
